EventListener/TagSubscriber.php

```diff
@@ -86,8 +113,13 @@ public function onKernelResponse(FilterResponseEvent $event)
         }
 
         if ($request->isMethodSafe()) {
-            // For safe requests (GET and HEAD), set cache tags on response
-            $this->cacheManager->tagResponse($response, $tags);
+            if ($event->isMasterRequest()) {
+                // For safe requests (GET and HEAD), set cache tags on response
+                $this->cacheManager->tagResponse($response, $tags);
+            } else {
+                // Retain tags to attach to response to master request
+                $this->tags = $tags;
+            }
```

Comment:
pretty much so, yes: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php#L492 - this is also the code path reached when esi is done (its only considered a sub request inside the HttpCache but then passed as a new master request to HttpKernel - which seems to me the expected and correct behaviour)